pkg/volume/csi/csi_block.go

```diff
@@ -331,31 +372,75 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
 		return errors.New(log("blockMapper.TearDownDevice failed to get CSI client: %v", err))
 	}
 
-	// Call NodeUnpublishVolume
-	publishPath := m.getPublishPath()
-	if _, err := os.Stat(publishPath); err != nil {
+	// Call NodeUnstageVolume
+	stagingPath := m.getStagingPath()
+	if _, err := os.Stat(stagingPath); err != nil {
 		if os.IsNotExist(err) {
-			klog.V(4).Infof(log("blockMapper.TearDownDevice publishPath(%s) has already been deleted, skip calling NodeUnpublishVolume", publishPath))
+			klog.V(4).Infof(log("blockMapper.TearDownDevice stagingPath(%s) has already been deleted, skip calling NodeUnstageVolume", stagingPath))
 		} else {
 			return err
 		}
 	} else {
-		err := m.unpublishVolumeForBlock(ctx, csiClient, publishPath)
+		err := m.unstageVolumeForBlock(ctx, csiClient, stagingPath)
 		if err != nil {
 			return err
 		}
 	}
 
-	// Call NodeUnstageVolume
-	stagingPath := m.getStagingPath()
-	if _, err := os.Stat(stagingPath); err != nil {
+	return nil
+}
+
+// UnmapPodDevice unmaps the block device path.
+func (m *csiBlockMapper) UnmapPodDevice(globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
+	if !m.plugin.blockEnabled {
+		return errors.New("CSIBlockVolume feature not enabled")
+	}
+
+	klog.V(4).Infof(log("unmapper.UnmapPodDevice(globalMapPath=%s, volumeMapPath,=%s, volumeMapName=%s, podUID=%s)",
+		globalMapPath, volumeMapPath, volumeMapName, string(podUID)))
+
+	err := ioutil.UnmapBlockVolume(globalMapPath, volumeMapPath, volumeMapName, podUID)
+	if err != nil {
+		return err
+	}
+
+	publishPath := m.getPublishPath()
+	blkUtil := volumepathhandler.NewBlockVolumePathHandler()
+	// Search under globalMapPath dir if all symbolic links from pods have been removed already.
+	// If symbolic links are there, pods may still refer the volume.
+	refs, err := blkUtil.GetDeviceSymlinkRefs(publishPath, globalMapPath)
+	if err != nil {
+		return err
+	}
+	if len(refs) > 0 {
+		return fmt.Errorf("blockMapper.UnmapPodDevice The device %q is still referenced from other Pods %v", globalMapPath, refs)
```

Comment:
this seems wrong, I thought we want to call NodeUnpublish every time we call UnmapDevice from the UnmapVolume operation.

Reply:
This is intended to avoid releasing descriptor lock by calling `DetachFileDevice`, if other pods are still using this device.

For CSI plugins that don't implement node stage/unstage, actual detach of the device from the node will happen in node unpublish. So, this kind of plugins requires descriptor lock to be released before calling node unpublish. 
(Releasing descriptor lock before calling `TearDownDevice`, or node unstage, in `GenerateUnmapDeviceFunc` is too late for such plugins. Actually, CSI rbd puglin doesn't work well with this code.)

And this code seems to be incorrect for CSI plugins that implement node stage/unstage, as you pointed out.

I started to think that we might need to move node unpublish to `TearDownDevice` if CSI plugins don't implement node stage/unstage. WDYT?

Reply:
If a CSI plugin is capable of publishing the same volume ID to two different paths (MULTI_NODE) it must also be capable of unpublishing the same volume ID from them independently. The current code makes it so that if I NodePublish(vol-a, path-1) and NodePublish(vol-a, path-2), only one of NodeUnpublish(vol-a, path-1) and NodeUnpublish(vol-a, path-2) will ever be called, but I would expect both to be called eventually, and that's why I say something seems wrong.

I understand this logic is here because the file descriptor lock will prevent, for certain plugins, a NodeUnpublish(vol-a, path-x) from succeeding. I would rather see us change the file descriptor lock logic than move the unpublish call, because I think the way you have aligned CSI calls to operations makes the most sense still.

What if we put this code in operation generator's UnmapVolume (to before UnmapPodDevice is called) and removed it from operation generator's UnmapDevice?

Reply:
Thank you for your suggestion. I've changed as suggested. 
By this change, we don't need any special handlings in CSI codes for it.

Also, as we changed publish path, or `devicePath`,  to be per pod for CSI, we will be able to expect only one symlink to the `devicePath` for CSI driver in `unmapVolumeFunc`.
(On the other hand, in-tree driver would stack here if the same block volume is attached to multiple pods on the same node, until all the pods on the same node will be deleted (Unless devicePath become per pod, too.))

PTAL

Reply:
Ummm... With this code, number of symlinks won't be decremented because `UnmapPodDevice` isn't called, when more than one symlinks exist, as a result `UnmapPodDevice` will never be called for the above in-tree situation.

We might be able to move `UnmapBlockVolume` to `unmapVolumeFunc`  from driver's `UnmapPodDevice` to work around this, but I'm not sure this will be the right direction.

Reply:
> We might be able to move UnmapBlockVolume to unmapVolumeFunc from driver's UnmapPodDevice to work around this, but I'm not sure this will be the right direction.

I think something like this is the correct fix. Right now, it is implicitly required that every plugin must execute this block of code to put a symlink in globalMapPath: 
https://github.com/kubernetes/kubernetes/blob/8635f7a0b09301227470ca7c229fc1aab5458984/pkg/volume/util/util.go#L521
and them remove it later:
https://github.com/kubernetes/kubernetes/pull/74026/files#diff-c3acf7a7fec401d0a3cb4998088ee1e5R547

Instead of having this implicit requirement, and every plugin doing the same thing, we should move this logic outside of the plugin and into the operation_generator. The plugin implementation should not have to worry about the file lock or the globalMapPath, if we keep that logic isolated to operation_generator I think it will be much easier to understand.

Reply:
Thank you for your comment. I will go ahead and fix further.
Also, I found that descriptor lock is taken every time after map, so we should release it every time before map. (And, it could be inside map/unmap utility function.)

Reply:
> Also, as we changed publish path, or devicePath, to be per pod for CSI, we will be able to expect only one symlink to the devicePath for CSI driver in unmapVolumeFunc.

Hmm I am having trouble following, which symlink do you mean? globalMapPath is still per-volume, not per-pod https://github.com/kubernetes/kubernetes/pull/74026/files#diff-ed824558531f8d1f616bc2f976266b64R53 and publishPath used to be per-volume, now it is per-pod. 

globalMapPath is per-volume and is for ref-counting via symlinks to publishPath/devicePath
publishPath is per-pod and is for 
- symlinking to pod mount path (e.g. /var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{outerVolumeSpecName}/) 
- locking (AttachFileDevice)

We lock the publishPath after CSI NodePublish.
a. if the publishPath is a symlink to another file, that file will be locked
b. if the publishPath is its own file, it will be locked

Now, let's say the same volume is NodePublished again to a different path.
a. if the publishPath is a symlink to another file, that file is already locked: no problem, idempotent
b. if the publishPath is its own file, it will be locked.

What do we do when one of the volume needs to be unmapped.
a. if the publishPath is a symlink to another file, we should not unlock it
b. if the publishPath is its own file, we should unlock it

How do we distinguish between case a and case b...

Reply:
To summarize issue, the in-tree plugins could always safely assume that there is a 1:1 mapping between  volumeID and devicePath. But now there could be 1:many mapping between volumeID and devicePath.

Reply:
also there is a case c where publishPath is a bind mount to a /dev/a. in which case I don't think there needs to be a lock on /dev/a at all because device will be busy and can't be detached.

I am not an expert on why the lock is here in the first place.

Reply:
I was only thinking about UnmapPodDevice call for CSI plugin is not blocked by current reference count logic. 

As you explained, we need to think about adding another logics:
  - To block calling node unpublish for CSI plugins that don't have unstage if the device is still in use,
  - To block calling node unstage for CSI plugins that have unstage if the device is still in use.

I've manually tested how loopback device works with bind mount and symlink. Please see below:

<details>

```
[Bind mount case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}
# touch /tmp/publish/pod1
# touch /tmp/publish/pod2

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# mount --bind /dev/loop0 /tmp/publish/pod1
# mount --bind /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
brw-rw----. 1 root disk 7, 0 Oct  3 20:50 pod1
brw-rw----. 1 root disk 7, 0 Oct  3 20:50 pod2

# ls -l /tmp/globalmap/
total 0
lrwxrwxrwx. 1 root root 17 Oct  3 20:52 pod1 -> /tmp/publish/pod1
lrwxrwxrwx. 1 root root 17 Oct  3 20:53 pod2 -> /tmp/publish/pod2

6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/publish/pod1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/publish/pod1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

# umount /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root    0 Oct  3 20:50 pod1
brw-rw----. 1 root disk 7, 0 Oct  3 21:14 pod2

# losetup -d /dev/loop2
# losetup 

# umount /tmp/publish/pod2
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root 0 Oct  3 20:50 pod1
-rw-r--r--. 1 root root 0 Oct  3 20:50 pod2

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/


[Symlink case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# ln -s /dev/loop0 /tmp/publish/pod1 
# ln -s /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod1 -> /dev/loop0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod2 -> /dev/loop0

# ls -l /tmp/globalmap/
total 0
lrwxrwxrwx. 1 root root 17 Oct  3 21:20 pod1 -> /tmp/publish/pod1
lrwxrwxrwx. 1 root root 17 Oct  3 21:20 pod2 -> /tmp/publish/pod2

6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /dev/loop0

# rm -f /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod2 -> /dev/loop0


# losetup -d /dev/loop2
# losetup 

# rm -f /tmp/publish/pod2
# ls -l /tmp/publish/
total 0

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/
```
</details>

Reply:
I'm thinking about making globalMap a bind mount instead of symlink and take descriptor lock for globalMap instead of devicePath (Then, it will be per pod lock, as well as ref count. Also, it will be idempotent).

If descriptor lock is taken to devicePath and CSI driver makes device available on publishPath as a symlink, we can't distinguish for which pod descriptor lock was taken, because losetup -j show the resolved path (Please see [Symlink case]'s step 6 in details sections of previous comment).

If we change globalMap to bind mount and take descriptor lock to it, we can see which descriptor lock is taken for which pod and who takes descriptor lock for the device (Please compare step 4 and step 6 in details sections of this comment and previous one).

[Current implementation]
````
                                    devicePath                            globalMap
                                   -------------                   ----------------------
1. actual device -- symlink    --> publishPath     <-- symlink --   glolbalMapPath/PodUID
2. actual device -- bind mount --> publishPath     <-- symlink --   glolbalMapPath/PodUID
3.                                 actual device   <-- symlink --   glolbalMapPath/PodUID
                                     ^--- descriptor lock
````

[This idea]
````
                                    devicePath                               globalMap
                                   -------------                       ----------------------
1. actual device -- symlink    --> publishPath     <-- bind mount --   glolbalMapPath/PodUID
2. actual device -- bind mount --> publishPath     <-- bind mount --   glolbalMapPath/PodUID
3.                                 actual device   <-- bind mount --   glolbalMapPath/PodUID
                                                                        ^--- descriptor lock
````

<details>

````
[Bind mount case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}
# touch /tmp/publish/pod1
# touch /tmp/publish/pod2

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# mount --bind /dev/loop0 /tmp/publish/pod1
# mount --bind /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# touch /tmp/globalmap/{pod1,pod2}
# mount --bind /tmp/publish/pod1 /tmp/globalmap/pod1
# mount --bind /tmp/publish/pod2 /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod2

# ls -l /tmp/globalmap/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod2

# findmnt | grep pod
├─/tmp/publish/pod1                   devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
├─/tmp/publish/pod2                   devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
├─/tmp/globalmap/pod1                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
└─/tmp/globalmap/pod2                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755

6. Take descriptor lock
# losetup /dev/loop1 /tmp/globalmap/pod1
# losetup /dev/loop2 /tmp/globalmap/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# losetup -j /tmp/globalmap/pod1
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

# losetup -j /tmp/globalmap/pod2
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2


# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# umount /tmp/globalmap/pod1
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:34 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:40 pod2

# umount /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:40 pod2

# losetup -d /dev/loop2
# losetup 

# umount /tmp/globalmap/pod2
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:34 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:34 pod2

# umount /tmp/publish/pod2
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:33 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:33 pod2

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/


[Symlink case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# ln -s /dev/loop0 /tmp/publish/pod1 
# ln -s /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

# touch /tmp/globalmap/{pod1,pod2}
# mount --bind /tmp/publish/pod1 /tmp/globalmap/pod1
# mount --bind /tmp/publish/pod2 /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod1 -> /dev/loop0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod2 -> /dev/loop0

# ls -l /tmp/globalmap/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:44 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:44 pod2

# findmnt | grep pod
├─/tmp/globalmap/pod1                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
└─/tmp/globalmap/pod2                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755


6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

# losetup /dev/loop1 /tmp/globalmap/pod1
# losetup /dev/loop2 /tmp/globalmap/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# losetup -j /tmp/globalmap/pod1
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

# losetup -j /tmp/globalmap/pod2
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)


7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2


# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /dev/loop0

# umount /tmp/globalmap/pod1
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:45 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:48 pod2

# rm -f /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod2 -> /dev/loop0

# losetup -d /dev/loop2
# losetup 

# umount /tmp/globalmap/pod2
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:45 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:45 pod2

# rm -f /tmp/publish/pod2
# ls -l /tmp/publish/
total 0

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/
````
</details>

Reply:
@wongma7 

Changed as commented above (Please also see the details below, for how pathes look like when block volumes are mapped). I will work on further debugging and improvement, but I would appreciate it if you could also review the current code.

<details>

```
Events:
  Type     Reason                  Age        From                     Message
  ----     ------                  ----       ----                     -------
  Normal   Scheduled               <unknown>  default-scheduler        Successfully assigned default/pod001-block to 127.0.0.1
  Normal   SuccessfulAttachVolume  25s        attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278"
  Warning  FailedMapVolume         16s        kubelet, 127.0.0.1       MapVolume.MapBlockVolume failed for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" : device not found
  Normal   SuccessfulMountVolume   15s        kubelet, 127.0.0.1       MapVolume.MapPodDevice succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev"
  Normal   SuccessfulMountVolume   15s        kubelet, 127.0.0.1       MapVolume.MapPodDevice succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" volumeMapPath "/var/lib/kubelet/pods/c34639f6-0a8f-458f-bf8e-24a284788187/volumeDevices/kubernetes.io~csi"
  Normal   Pulled                  14s        kubelet, 127.0.0.1       Container image "k8s.gcr.io/busybox:1.24" already present on machine
  Normal   Created                 13s        kubelet, 127.0.0.1       Created container container001
  Normal   Started                 12s        kubelet, 127.0.0.1       Started container container001


# ls -l /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev
total 0
brwxrwxrwx. 1 root disk 251, 80 Oct  5 00:37 c34639f6-0a8f-458f-bf8e-24a284788187

# ls -l /var/lib/kubelet/pods/c34639f6-0a8f-458f-bf8e-24a284788187/volumeDevices/kubernetes.io~csi
total 4
lrwxrwxrwx. 1 root root 175 Oct  5 00:37 pvc-82051086-eb98-4dd0-941a-1512bc438278 -> /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/0001-0009-rook-ceph-0000000000000001-3745bc30-e708-11e9-b798-0242ac110005/c34639f6-0a8f-458f-bf8e-24a284788187

# losetup | grep '/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev'
/dev/loop7         0      0         0  0 /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187                                        0

# findmnt /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187
TARGET                                                                                                                                     SOURCE          FSTYPE   OPTIONS
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187 devtmpfs[/rbd5] devtmpfs rw,seclabel,size=16460544

# lsblk | grep rbd5
rbd5   251:80   0    1G  0 disk 

# mount | grep /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187
devtmpfs on /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187 type devtmpfs (rw,seclabel,size=16460544k,nr_inodes=4115136,mode=755)
```

</details>

Reply:
I am curious about the FailedMapVolume event shown above.
Did you find out the cause of it ?

Thanks

Reply:
Thank you for pointing it out.

I've found that this is due to the difference of losetup command's output depending on the arguments. I've fixed it in the latest commit. PTAL

(In the previous code, `AttachFileDevice` failed because `makeLoopDevice` failed to find the device just created.)